### DISCUSSION TITLE
content/dev/afreecatv-crawling/

### DISCUSSION THREAD

**Discussion by giscus** [cha2hyun/cha2hyun.github.io#87] [Comments]
# content/projects/%EB%B0%B0%EB%8F%8C%EC%9D%B4%EC%9D%98%EB%8B%B9%EA%B5%AC%EC%83%9D%ED%99%9C/afreecatv-crawling/

리버스엔지니어링 + 소켓연결 + 네트워크탭 분석. feat) 아프리카도우미

https://cha2hyun.blog/content/projects/%EB%B0%B0%EB%8F%8C%EC%9D%B4%EC%9D%98%EB%8B%B9%EA%B5%AC%EC%83%9D%ED%99%9C/afreecatv-crawling/

**Comment by DOCHIS** [comment]
좋은 글 감사합니다 코드 관련으로 아직 해석 방법이 확실하지 않아 찾아보고는 있는데 5분 마다 끊기는 현상은 인터벌을 통해 000000000100 코드를 쏘시면 해결됩니다 해당 코드가 PING 코드로 보입니다

**Reply by cha2hyun** [reply]
게시글에 관심가져주시고 좋은 내용 공유 주셔서 감사합니다 👍

**Reply by DOCHIS** [reply]
한가지 더 알아낸 점이 있어요 ``000100058200`` 와 같은 코드 부분인데요.
코드는 12글자의 고정 길이로 반환이 되더라구요
.
소스코드를 뜯다보니 알아낸건데 이 부분은
- ChatServiceType 4자리
- 메시지의 길이 6자리
- 여분(추측) 2자리
로 구성되어있는것 같아요

먼저 ChatServiceType 부분은 제가 소스코드를 따서 [이 문서에](https://gist.github.com/DOCHIS/8095eb6a05586de220c81503b4684b36) 정리해 보았는데요 해당 코드를 각 상황에 맞게 사용하는것 같습니다

두번째로 메시지의 길이 부분인데요, 보낼 명령어를 Uint8Array로 변환 한 후 해당 문자열의 byteLength를 계산해서 해당 길이를 사용하더라구요 BJ의 아이디길이에 따라 달라진 점도 이 때문으로 보입니다

``두번째 핸드쉐이크``라고 표시해주신 패킷은 JOIN 패킷인데 이 패킷에는 ``CHATNO``와 ``FTK`` 정보가 포함되어있는데, ``FTK`` 정보에는 ``3152bc4e775a25a9d49928f61bf3e351_gurm01_255592134_html5_0`` 와 같이 BJ의 아이디 정보가 포함되어있습니다
때문에 정리해주신 데로 BJ의 아이디 길이에 따라 byteLength가 달라지기 때문에 변경되는점도 있고 ``CHATNO``도 상황에 따라 길이가 달라질 수 있기 때문에 말씀 주신 예외 케이스가 발생되고 있는것으로 보입니다

**Reply by cha2hyun** [reply]
감사합니다 :)

알려주신 내용 바탕으로 개선하였습니다.

**Comment by ppk052** [comment]
그대로 복붙하고 아프리카tv url입력하면 ERROR: no close frame received or sent가 계속떠요!
indent 문제 없어요! problems에 아무것도 안떠요...

**Reply by ppk052** [reply]
보니까 비쥬얼 스튜디오 코드에 붙여넣기 하니까 핸드쉐이크에서 특수문자부분에 esc ff이런게 생겨있네요 이게 원인인거 같은데 혹시 이거 어떻게 해결하는지 아시나요?...

**Reply by cha2hyun** [reply]
안녕하세요 @ppk052 님! 

JOIN 패킷의 예외적인 부분 때문에 진행되지 않았던 것 같습니다. 해당 부분을 보완하여 글을 수정했습니다. 참고해주시기 바랍니다.

**Comment by ilsun86** [comment]
좋은글 감사합니다 다만 3월15일자 보면 from api import get_player_live 했을때 get_player_live 부분을 불러올수 없다고 계속 뜨는데 해결방법을 알수있나요?

**Reply by cha2hyun** [reply]
에고 답변이 늦었습니다 해당 내용은 깃헙 레포에서 확인 하실 수 있어요 !

**Comment by CHOIMONK** [comment]
안녕하세요. 정말 도움이 됩니다 
저는 코딩 초보로 입문하고 있는 사람입니다
다름이 아니라 해당 아프리카TV 채팅 추출을 하려면 API 취득이 필요한데 
오로지 링크로만 실시간 댓글을 추출하는 방안은 없을까요 
유튜브 라이브 실시간 또한 추출되나 아프리카는 어렵네요 ㅠ

**Reply by cha2hyun** [reply]
@CHOIMONK 
안녕하세요 ! 해당 내용에 대한 소스코드는 전부 공개되어있습니다.
https://github.com/cha2hyun/afreecatv-chat-crawler

API 키를 취득하시거나 그러실 필요가 없습니다 ! 
https://github.com/cha2hyun/afreecatv-chat-crawler/blob/main/api.py

링크만으로 댓글 추출이 가능하므로 코드 보시고 궁금하신 점 있으시면 말씀 주세요 !

**Comment by im-pearl** [comment]
너무 많은 도움 되었습니다! 정말 감사드립니다! 진짜 대단하시네요..

**Reply by cha2hyun** [reply]
감사합니다 ㅎㅎ

**Comment by hamsterb0t** [comment]
안녕하세요, 좋은 글 감사합니다.
올려주신 코드를 채팅 유틸리티 개발에 사용하고자 하는데 괜찮을까요?

**Reply by hamsterb0t** [reply]
참고로, 개발 중인 소스코드는 비공개로 두고 있습니다. 유틸리티도 제한적으로 배포되고 있고, 출처는 유틸리티 내에 표기해 두려고 합니다.
감사합니다!

**Reply by cha2hyun** [reply]
@ywj515 
안녕하세요, 위 내용은 제가 개인적으로 학습 목적으로 작성한 것으로 출처 남겨주시고 사용해주시는 건 무방하나, 혹여나 상업적으로 사용하시거나 다른 용도로 활용하실때 발생하는 문제에 대해서는 제가 책임질 수 없음을 양해해주시기 바랍니다.

여쭤봐주셔서 감사합니다. 유틸리티 개발에 도움이 되길 바랍니다!

**Reply by hamsterb0t** [reply]
네, 취미로 개발하는 것이라서 상업적인 용도로 사용하지 않을 것이라 걱정하지 않으셔도 될 것 같습니다.허락해주셔서 감사합니다!

**Comment by Hoon1021** [comment]
접속과 관련하여 여쭤볼게 있습니다

messages 배열에서 
messages[0]이 0004000 으로 시작하고
messages[1] == "-1" messages[4] == "2" 일때
강제퇴장으로 칭하도록 하고, messages[5]가 1일때
BJ로 부터 강제퇴장, 2일때 매니저로 부터 강제퇴장으로
로그를 남기게 만들었는데,

1일때는 처리자를 BJID로 하면 되는데
2일때 매니저 아이디를 받아오려면 어떻게 해야할까요?

**Reply by purr729** [reply]
혹시 찾으셨나요..?

**Comment by eeyore0424** [comment]
안녕하세요 !
코딩 공부를 하고 있는 사람입니다 ! 이렇게 분석하고, 코드를 작성한다는게 정말 신기하네요.
혹시 현재는 아프리카티비에서 숲으로 바뀐거 같은데 해당 코드를 그대로 사용해도 채팅창 크롤링이 가능한건가요 ?! 아프리카 TV에서 제공되는 api 부분을 보니 url이 아프리카로 되어있어서 궁금해서 여쭤보아요 !

**Reply by eeyore0424** [reply]
아 그리고 힘들게 분석하신거 이렇게 공유해주셔서 감사합니다. 새해 복 많이 받으세요 !

**Reply by ykwak22** [reply]
저도 궁금합니다. soop으로 바뀌면서 단순 url 뿐만 아니라 전반적인 코드가 다 바뀌었나요?

**Reply by Rrrriaa** [reply]
그대로 써도 정상 작동합니다! api.py 에서 주소를 live.sooplive.co.kr로 바꿔도 작동하고 아무거나 써도 되네요

**Reply by cha2hyun** [reply]
@Rrrriaa 확인해주셔서 감사합니다 👍🏻

**Comment by mono1ith** [comment]
안녕하세요 포스팅 잘 봤습니다!
덕분에 원하는 프로그램을 만들 수 있었는데,
출처 남기고 개인 깃허브에 업로드 해도 괜찮을까요? (상업용 X)

**Reply by cha2hyun** [reply]
안녕하세요 답변이 늦었습니다. 
상업용 아니라면 자유롭게 출처 남기고 사용가능합니다.

감사합니다 !

**Comment by power3247** [comment]
이제  입장 퇴장 정보는 감춘것 같네요
